### ISSUE TITLE
Question regardless blocked signals in threads

### ISSUE THREAD

**Issue by azat** [confluentinc/librdkafka#4571]
I've recently came across this (https://github.com/ClickHouse/ClickHouse/pull/57907), what is the problem with signals? Can you please clarify on this?

**Comment by emasab** [comment]
The signal mask is set to block all signals. A new thread is created that inherits the signal mask blocking all signals. Then the original signal mask is restored. This avoids that signals handlers are executed in librdkafka threads.
A safer way would be to set `pthread_attr_setsigmask_np` while creating the thread but this isn't portable (_np), only supported by GNU pthreads.

**Comment by emasab** [comment]
It restores the original signmask, not the one with all blocked signal,  check these lines
https://github.com/confluentinc/librdkafka/blob/62165e6d09082454138948922b96b54bd55058a8/rdkafka.c#L922
https://github.com/confluentinc/librdkafka/blob/62165e6d09082454138948922b96b54bd55058a8/rdkafka.c#L950

The race condition Magnus was referring to is happens if the signal handler is executed in the librdkafka thread instead of the application thread it was meant to run, that's why signal are masked in the created librdkafka thread.

**Comment by emasab** [comment]
> And while creating it also blocks signals but then restore it to all-blocked, since they had been blocked already above:
No, they are restored to `oldset`, that is the value they had before calling `pthread_sigmask`, it could be any value you had set, not all blocked.

**Comment by azat** [comment]
Yes, you are talking about `pthread_sigmask` behavior, but the problem is that it already inherited the all blocked sigmask.

```
rd_kafka_new() -> blocks all signals -> rd_kafka_broker_add() ->
    get old sigmask (all signals blocked) -> create thread -> restore sigmask
restore sigmask
```

So I mean't at least something like this (though not sure that this is the only place):

```patch
diff --git a/src/rdkafka.c b/src/rdkafka.c
index e30faa6c..0138dfe3 100644
--- a/src/rdkafka.c
+++ b/src/rdkafka.c
@@ -2417,12 +2417,6 @@ rd_kafka_t *rd_kafka_new (rd_kafka_type_t type, rd_kafka_conf_t *app_conf,
          * @warning `goto fail` is prohibited past this point
          */

-        rdk_thread_mutex_lock(&rk->rk_internal_rkb_lock);
-       rk->rk_internal_rkb = rd_kafka_broker_add(rk, RD_KAFKA_INTERNAL,
-                                                 RD_KAFKA_PROTO_PLAINTEXT,
-                                                 "", 0, RD_KAFKA_NODEID_UA);
-        rdk_thread_mutex_unlock(&rk->rk_internal_rkb_lock);
-
        /* Add initial list of brokers from configuration */
        if (rk->rk_conf.brokerlist) {
                if (rd_kafka_brokers_add0(rk, rk->rk_conf.brokerlist) == 0)
@@ -2430,6 +2424,12 @@ rd_kafka_t *rd_kafka_new (rd_kafka_type_t type, rd_kafka_conf_t *app_conf,
                                        "No brokers configured");
        }

+        rdk_thread_mutex_lock(&rk->rk_internal_rkb_lock);
+       rk->rk_internal_rkb = rd_kafka_broker_add(rk, RD_KAFKA_INTERNAL,
+                                                 RD_KAFKA_PROTO_PLAINTEXT,
+                                                 "", 0, RD_KAFKA_NODEID_UA);
+        rdk_thread_mutex_unlock(&rk->rk_internal_rkb_lock);
+
 #ifndef _WIN32
        /* Restore sigmask of caller */
        pthread_sigmask(SIG_SETMASK, &oldset, NULL);
```

**Comment by felixge** [comment]
The current behavior of blocking all signals is also unfortunate for runtime CPU profilers. For example, Go applications using librdkafka can't use Go's built-in CPU profiler to understand the CPU usage from the kafka library because the Go runtime uses `setitimer` and `timer_create` to capture stack traces after the process (or a thread) has used up a certain amount of CPU time. See this [blog post](https://www.datadoghq.com/blog/engineering/profiling-improvements-in-go-1-18/) for more details.

**Comment by weissi** [comment]
Indeed, this is a problem. I filed #5255 with a full reproducer that shows the Kafka threads with everything blocked too in case that helps.
